File: src/resolver/paths.ts
import * as os from "node:os";
import * as path from "node:path";

import { InvalidManifestError } from "../errors/index.js";

import type { IncludeRef, IncludeSourceKind } from "./types.js";

export interface ResolverPaths {
  /** Absolute project root. */
  projectRoot: string;
  /** Absolute path to .claude-profiles. */
  profilesDir: string;
  /** Absolute path to .claude-profiles/_components. */
  componentsDir: string;
}

export function buildPaths(projectRoot: string): ResolverPaths {
  const root = path.resolve(projectRoot);
  return {
    projectRoot: root,
    profilesDir: path.join(root, ".claude-profiles"),
    componentsDir: path.join(root, ".claude-profiles", "_components"),
  };
}

export function profileDir(paths: ResolverPaths, name: string): string {
  return path.join(paths.profilesDir, name);
}

/**
 * Windows-reserved DOS device names. Rejected on every host OS so a profile
 * authored on Linux/macOS cannot land on Windows under a name the kernel
 * refuses to open. Match is case-insensitive and includes any extension
 * (CONFIG.SYS-style "CON.txt" → CON-the-device on Windows).
 *
 * Source: CONTRIBUTING.md "Reserved filenames" + R39 cross-platform contract.
 */
const WIN_RESERVED_NAMES = /^(CON|PRN|AUX|NUL|COM[1-9]|LPT[1-9])(\..*)?$/i;

/**
 * R2: a profile identifier is the bare directory name under
 * `.claude-profiles/`. Reject anything that could escape that boundary
 * (slashes, backslashes, parent-traversal, leading `.` or `_` per the
 * `_components`/hidden-dir convention enforced by listProfiles, empty).
 *
 * Cross-platform safety (R39 / CONTRIBUTING.md): also reject names that
 * Windows's filesystem refuses — the DOS-device names CON/PRN/AUX/NUL/
 * COM1-9/LPT1-9 (with or without an extension), and trailing dot or space
 * (silently stripped by Win32 → name collisions). Enforced on every host
 * so a profile authored on Linux/macOS remains valid on Windows.
 *
 * Predicate; callers convert a `false` result into a CliUserError or
 * MissingProfileError depending on context.
 */
export function isValidProfileName(name: string): boolean {
  if (name.length === 0) return false;
  if (name === "." || name === "..") return false;
  if (name.startsWith("_") || name.startsWith(".")) return false;
  if (name.includes("/") || name.includes("\\") || name.includes(path.sep)) return false;
  if (name.includes("\0")) return false;
  if (WIN_RESERVED_NAMES.test(name)) return false;
  // Win32 silently strips trailing dot and space → 'foo.' and 'foo ' both
  // resolve to 'foo'. Reject so a profile authored on POSIX cannot collide
  // with another profile on Windows after path normalisation.
  if (name.endsWith(".") || name.endsWith(" ")) return false;
  return true;
}

/**
 * True iff `name` matches a Windows reserved DOS-device name. Exported for
 * defense-in-depth callers (state/paths.ts) that re-validate at the persist
 * boundary. Predicate-only; callers shape the error themselves.
 */
export function isWindowsReservedName(name: string): boolean {
  return WIN_RESERVED_NAMES.test(name);
}

/**
 * Classify and resolve an includes entry per R37. The four valid forms:
 *  - `bare-name`         → component (rooted at `_components/<name>/`)
 *  - `./...` or `../...` → relative (resolved from the referencing profile dir)
 *  - `/...`              → absolute
 *  - `~/...` or `~`      → tilde (expanded against $HOME)
 *
 * Anything else — notably bare strings containing `/` or platform separators
 * that aren't `./`/`../`/absolute — is rejected via InvalidManifestError.
 *
 * `external` is the orthogonal semantic flag: true if the resolved path
 * falls outside the project root. `component` is always non-external (rooted
 * under `_components/`). `absolute`, `tilde`, and `relative` may be external
 * — for example, `./../../outside-project` is a valid relative form that
 * resolves outside the project root and is correctly flagged.
 */
export function classifyInclude(
  raw: string,
  referencingProfileDir: string,
  paths: ResolverPaths,
  referencedBy: string,
): IncludeRef {
  let kind: IncludeSourceKind;
  let resolvedPath: string;

  if (raw.length === 0) {
    throw new InvalidManifestError(
      referencingProfileDir,
      `empty include string in profile "${referencedBy}"`,
    );
  }

  // `.` and `..` are not legal R37 forms — they have no `./`/`../` prefix
  // so they would otherwise fall through to the bare-component branch and
  // resolve to `_components/` itself or `.claude-profiles/`, neither of
  // which is the author's intent. Reject up-front.
  if (raw === "." || raw === "..") {
    throw new InvalidManifestError(
      referencingProfileDir,
      `include "${raw}" in profile "${referencedBy}" is not a valid form — use a bare component name, "./..." / "../..." for relative, "/..." for absolute, or "~/..." for home-relative`,
    );
  }

  if (raw === "~" || raw.startsWith("~/")) {
    const home = os.homedir();
    const rest = raw === "~" ? "" : raw.slice(2); // strip "~/"
    resolvedPath = rest === "" ? path.resolve(home) : path.resolve(home, rest);
    kind = "tilde";
  } else if (raw.startsWith("~")) {
    // "~user" form: not portable, not in R37. Reject.
    throw new InvalidManifestError(
      referencingProfileDir,
      `include "${raw}" in profile "${referencedBy}" — "~user" form is not supported; use "~/path" or an absolute path`,
    );
  } else if (path.isAbsolute(raw)) {
    resolvedPath = path.resolve(raw);
    kind = "absolute";
  } else if (raw.startsWith("./") || raw.startsWith("../")) {
    resolvedPath = path.resolve(referencingProfileDir, raw);
    kind = "relative";
  } else if (raw.includes("/") || raw.includes("\\") || raw.includes(path.sep)) {
    // R37 admits exactly four forms; a bare-with-slashes string matches
    // none of them. Reject explicitly rather than silently routing. We
    // reject backslashes on every platform so a manifest authored on
    // Windows behaves identically on Linux/macOS.
    throw new InvalidManifestError(
      referencingProfileDir,
      `include "${raw}" in profile "${referencedBy}" is not a valid form — use a bare component name, "./..." / "../..." for relative, "/..." for absolute, or "~/..." for home-relative`,
    );
  } else {
    resolvedPath = path.join(paths.componentsDir, raw);
    kind = "component";
  }

  const external = isExternal(resolvedPath, paths.projectRoot);
  return { raw, kind, resolvedPath, external };
}

export function isExternal(absPath: string, projectRoot: string): boolean {
  const rel = path.relative(projectRoot, absPath);
  if (rel === "") return false;
  // Match only `..` (the parent itself) or a `..` segment followed by the
  // platform separator. A bare filename like `..hidden` whose first
  // characters happen to be `..` is a legitimate in-root path, not a
  // traversal segment — `rel.startsWith("..")` would have wrongly flagged
  // it external.
  if (rel === ".." || rel.startsWith(".." + path.sep)) return true;
  return path.isAbsolute(rel);
}

File: go/internal/resolver/paths.go
package resolver

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	pipelineerrors "github.com/htxryan/c3p/internal/errors"
)

// ResolverPaths holds the canonical paths derived from a project root.
type ResolverPaths struct {
	ProjectRoot   string
	ProfilesDir   string
	ComponentsDir string
}

// BuildPaths constructs ResolverPaths rooted at projectRoot. The input is
// canonicalized via filepath.Abs so callers may pass relative paths.
func BuildPaths(projectRoot string) ResolverPaths {
	root, err := filepath.Abs(projectRoot)
	if err != nil {
		// filepath.Abs only fails when os.Getwd fails; fall back to the input
		// rather than panicking — tests construct paths from t.TempDir() which
		// is already absolute.
		root = projectRoot
	}
	return ResolverPaths{
		ProjectRoot:   root,
		ProfilesDir:   filepath.Join(root, ".claude-profiles"),
		ComponentsDir: filepath.Join(root, ".claude-profiles", "_components"),
	}
}

// ProfileDir returns the absolute directory for a profile by name.
func ProfileDir(paths ResolverPaths, name string) string {
	return filepath.Join(paths.ProfilesDir, name)
}

// winReservedNames matches Windows DOS-device names with optional extension.
// Rejected on every host so a profile authored on POSIX cannot land on
// Windows under a name the kernel refuses to open.
var winReservedNames = regexp.MustCompile(`(?i)^(CON|PRN|AUX|NUL|COM[1-9]|LPT[1-9])(\..*)?$`)

// IsValidProfileName returns true iff name is a legal profile identifier
// (R2 + R39 + PR16 cross-platform safety).
func IsValidProfileName(name string) bool {
	if name == "" {
		return false
	}
	if name == "." || name == ".." {
		return false
	}
	if strings.HasPrefix(name, "_") || strings.HasPrefix(name, ".") {
		return false
	}
	// Both POSIX and Windows separators rejected on every host so a manifest
	// authored on either platform behaves identically.
	if strings.ContainsAny(name, "/\\") {
		return false
	}
	if strings.ContainsRune(name, 0) {
		return false
	}
	if winReservedNames.MatchString(name) {
		return false
	}
	if strings.HasSuffix(name, ".") || strings.HasSuffix(name, " ") {
		return false
	}
	return true
}

// IsWindowsReservedName returns true iff name matches a Windows DOS-device
// reserved name (PR16). Exported for defense-in-depth callers that re-validate
// at the persist boundary.
func IsWindowsReservedName(name string) bool {
	return winReservedNames.MatchString(name)
}

// IsExternal returns true iff absPath is outside projectRoot. Both arguments
// must already be absolute; the function does not canonicalize and does not
// follow symlinks (textual containment only — security boundary doc note).
func IsExternal(absPath, projectRoot string) bool {
	rel, err := filepath.Rel(projectRoot, absPath)
	if err != nil {
		return true
	}
	if rel == "." {
		return false
	}
	// `..` alone, or `../something`, signals escape. Guard against a bare
	// filename like `..hidden` (legitimate) being mistaken for traversal —
	// match only `..` or a `..` segment followed by the path separator.
	if rel == ".." || strings.HasPrefix(rel, ".."+string(filepath.Separator)) {
		return true
	}
	if filepath.IsAbs(rel) {
		return true
	}
	return false
}

// ClassifyInclude classifies and resolves an includes entry per R37 +
// PR16a. Returns InvalidManifestError for syntactically invalid forms,
// PathTraversalError for relative-form (`./` / `../`) entries that resolve
// outside projectRoot.
//
// PR16a is enforced via TEXTUAL canonicalization (filepath.Join) only —
// the function does NOT follow symlinks. A relative include that points at
// a symlink whose target lives outside projectRoot will pass classification
// (the textual path is inside) and the link will be followed by os.Stat at
// resolve time. v1 accepts that boundary; D5 (state) and D6 (drift) trust
// the resolver's classification.
//
// TS deviation note: the TS resolver tracks `../`-outside-projectRoot
// includes as `external=true` and surfaces them as MissingInclude only
// when the directory does not exist. The Go resolver rejects up-front with
// PathTraversalError per the spec strengthening in PR16a. Downstream
// consumers (D2/D5/D7) will therefore never see a relative-kind include
// with `external=true` on Go output.
func ClassifyInclude(raw, referencingProfileDir string, paths ResolverPaths, referencedBy string) (IncludeRef, error) {
	if raw == "" {
		return IncludeRef{}, pipelineerrors.NewInvalidManifestError(
			referencingProfileDir,
			fmt.Sprintf("empty include string in profile %q", referencedBy),
		)
	}

	var (
		kind         IncludeSourceKind
		resolvedPath string
	)

	switch {
	case raw == "~" || strings.HasPrefix(raw, "~/"):
		home, err := os.UserHomeDir()
		if err != nil {
			return IncludeRef{}, pipelineerrors.NewInvalidManifestError(
				referencingProfileDir,
				fmt.Sprintf("cannot expand %q in profile %q: home directory unavailable: %v", raw, referencedBy, err),
			)
		}
		if raw == "~" {
			resolvedPath = filepath.Clean(home)
		} else {
			resolvedPath = filepath.Join(home, raw[2:])
		}
		// Make absolute (Clean does not).
		resolvedPath = mustAbs(resolvedPath)
		kind = IncludeKindTilde

	case strings.HasPrefix(raw, "~"):
		// "~user" form: not portable, not in R37.
		return IncludeRef{}, pipelineerrors.NewInvalidManifestError(
			referencingProfileDir,
			fmt.Sprintf(
				`include %q in profile %q — "~user" form is not supported; use "~/path" or an absolute path`,
				raw, referencedBy,
			),
		)

	case filepath.IsAbs(raw):
		resolvedPath = filepath.Clean(raw)
		kind = IncludeKindAbsolute

	case strings.HasPrefix(raw, "./") || strings.HasPrefix(raw, "../"):
		resolvedPath = filepath.Join(referencingProfileDir, raw)
		// PR16a: a relative include with `../` segments that escapes the
		// project root must be rejected before any filesystem touch.
		if IsExternal(resolvedPath, paths.ProjectRoot) {
			return IncludeRef{}, pipelineerrors.NewPathTraversalError(raw, resolvedPath, referencedBy)
		}
		kind = IncludeKindRelative

	case strings.ContainsAny(raw, "/\\") || strings.ContainsRune(raw, os.PathSeparator):
		// R37 admits exactly four forms; bare-with-slashes matches none.
		// Reject explicitly so Windows-authored manifests behave identically.
		return IncludeRef{}, pipelineerrors.NewInvalidManifestError(
			referencingProfileDir,
			fmt.Sprintf(
				`include %q in profile %q is not a valid form — use a bare component name, "./..." / "../..." for relative, "/..." for absolute, or "~/..." for home-relative`,
				raw, referencedBy,
			),
		)

	default:
		resolvedPath = filepath.Join(paths.ComponentsDir, raw)
		kind = IncludeKindComponent
	}

	return IncludeRef{
		Raw:          raw,
		Kind:         kind,
		ResolvedPath: resolvedPath,
		External:     IsExternal(resolvedPath, paths.ProjectRoot),
	}, nil
}

func mustAbs(p string) string {
	if filepath.IsAbs(p) {
		return filepath.Clean(p)
	}
	abs, err := filepath.Abs(p)
	if err != nil {
		return filepath.Clean(p)
	}
	return abs
}
